### .Imports

In [2]:
import os
import pandas as pd
import json
import matplotlib.pyplot as plt
import numpy as np
import re
import math
import numpy as np

%pylab inline
pylab.rcParams['figure.figsize'] = (20,12)

Populating the interactive namespace from numpy and matplotlib


### .Constants

In [3]:
DATA_FOLDER = 'data_clean'
IS_DATE = re.compile("^[1-2]{1}[0-9]{3}$")

### .Usefull functions

#### Function to restrict the dataframe to a fix period in term of years

In [4]:
def get_period(df, start=None,end=None):
    columns_to_keep = []
    dates = []
    for elem in df.columns:
        if not IS_DATE.match(elem):
            columns_to_keep.append(elem)
        else:
            dates.append(int(elem))
    if start == None:
        start = np.min(dates)
    if end == None:
        end = np.max(dates)
    
    for date in dates:
        if (date<=end) & (date>=start):
            columns_to_keep.append(str(date))
    
    return df[columns_to_keep]

#### Function to get the age corresponding to the intervales as input

In [5]:
def build_age(intervales):
    ages_total = []
    for i in range(len(intervales)-1):
        ages = []
        ages.append(intervales[i])
        for k in range(int(intervales[i]),int(intervales[i+1])):
            if(k != 1 and k<100):
                ages.append('{} ans'.format(k))
            if(k == 1):
                ages.append('{} an'.format(k))
            if(k==100):
                ages.append('100 ans ou plus')
                break
        ages_total.append(ages)
    return ages_total

#### Function to describe a dataframe

In [6]:
def describeDf(df,name="DESCRIPTION",level=1):
    print("______________________________ " + name+" ______________________________\n")
    date = []
    if(level==1):
        print("\n")
        for col in df.columns:
            if((not IS_DATE.match(col)) & (not re.compile("[0-9]+ an").match(col))):
                print("         ATTRIBUTE: "+col)
                print("   "+str(df[col].unique()))
                print("\n")
            else:
                date.append(col)
        print("         VALUES: "+str(date))
        print("\n\n\n")
    else :
        print(df.columns.values)
        print("\n\n\n")

### .Load Dataframe

In [9]:
print("Dataframes available:\n\n")
for file in os.listdir(DATA_FOLDER):
    filename = os.fsdecode(file)
    tablename = "df_"+os.path.splitext(os.path.splitext(filename)[0])[0]
    globals()[tablename] = pd.read_csv(DATA_FOLDER + '/' + filename)
    try:
        globals()[tablename] = globals()[tablename].drop(columns=['Unnamed: 0'])
    except:
        pass
    describeDf(globals()[tablename], name=tablename,level=0)
    

Dataframes available:


______________________________ df_Qualite_vie_agglo ______________________________

['agglo_index' 'agglo' 'type_index' 'type' '2011' '2012' '2013' '2014']




______________________________ df_Population_age_1992 ______________________________

['age' 'sexe' '1992' '1993' '1994' '1995' '1996' '1997' '1998' '1999'
 '2000' '2001' '2002' '2003' '2004' '2005' '2006' '2007' '2008' '2009'
 '2010']




______________________________ df_Accident_cantons ______________________________

['canton_index' 'canton' 'gravite_index' 'gravite' 'ruralite_index'
 'ruralite' '1992' '1993' '1994' '1995' '1996' '1997' '1998' '1999' '2000'
 '2001' '2002' '2003' '2004' '2005' '2006' '2007' '2008' '2009' '2010'
 '2011' '2012' '2013' '2014' '2015' '2016' '2017']




______________________________ df_Frais_routes_cantonales ______________________________

['annee' 'canton' 'Total' 'Constructions' 'Amélio-' 'Acquisition'
 'Entretien Construction' 'Entretien Exploitation' 'Administration'


In [10]:
describeDf(df_Frais_routes_cantonales)

______________________________ DESCRIPTION ______________________________



         ATTRIBUTE: annee
   [1994 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004 2005 2006 2007
 2008 2009 2010 2011 2012 2013 2014 2015]


         ATTRIBUTE: canton
   ['Appenzell Rh.-Ext.' 'Appenzell Rh.-Int.' 'Argovie' 'Berne'
 'Bâle-Campagne' 'Bâle-Ville' 'Espace Mittelland' 'Fribourg' 'Genève'
 'Glaris' 'Grisons' 'Jura' 'Lucerne' 'Neuchâtel' 'Nidwald' 'Obwald'
 'Région lémanique' 'Schaffhouse' 'Schwytz' 'Soleure' 'St. Gall'
 'Suisse centrale' 'Suisse du Nord-Ouest' 'Tessin' 'Thurgovie' 'Total'
 'Uri' 'Valais' 'Vaud' 'Zentralschweiz' 'Zoug' 'Zurich']


         ATTRIBUTE: Total
   [  31374.    4206.  100621.  169980.   88815.   99176.  392484.   50551.
   63226.   14798.  195564.   22143.   66472.  107991.    7941.    4959.
  323508.    7701.   27276.   41819.   81012.  391618.  288612.  218586.
   56963. 1907803.   38102.  139031.  121251.  176670.   31920.  116326.
   32061.    4552.   94320.  18338

## A. Test on the dataset

In this section, unknown values were not excluded to avoid asymetric operations on different datasets. 

In [13]:
reg = re.compile('201[1-7]')

#test dataset with cantons
df_acc_cantons = df_Accident_cantons

#only taking years from 2010 to 2017
toDrop = []
for elem in df_acc_cantons.columns:
    if reg.match(elem):
        toDrop.append(elem)
df_acc_cantons = df_acc_cantons.drop(columns=toDrop)

#aggregate groups by gravite 
acc_cantons = df_acc_cantons.groupby(['gravite']).sum().sort_values(['2010'], ascending=False)\
            [[col for col in df_acc_cantons.columns if IS_DATE.match(col)]]
display(acc_cantons)

1992     1993     1994     1995     1996  \
gravite                                                                      
Accidents avec blessés légers  15755.0  15933.0  16967.0  16409.0  15504.0   
Accidents avec blessés graves   6769.0   6258.0   5923.0   5981.0   5512.0   
Accidents mortels                748.0    661.0    637.0    640.0    562.0   

                                  1997     1998     1999     2000     2001  \
gravite                                                                      
Accidents avec blessés légers  16084.0  16207.0  17311.0  17754.0  17942.0   
Accidents avec blessés graves   5438.0   5475.0   5574.0   5428.0   5458.0   
Accidents mortels                553.0    550.0    549.0    555.0    496.0   

                                  2002     2003     2004     2005     2006  \
gravite                                                                      
Accidents avec blessés légers  17866.0  18168.0  17501.0  16719.0  16543.0   
Accidents avec blessés graves   5305.0   5173.0   4912.0   4614.0   4608.0   
Accidents mortels                476.0    499.0    478.0    373.0    340.0   

                                  2007     2008     2009     2010  
gravite                                                            
Accidents avec blessés légers  16790.0  16034.0  15834.0  15214.0  
Accidents avec blessés graves   4760.0   4366.0   4337.0   4082.0  
Accidents mortels                361.0    336.0    335.0    313.0

In [14]:
#test dataset with circonstances
df_acc_circo = df_Accident_circonstances

#only taking years from 2010 to 2017
toDrop = []
for elem in df_acc_circo.columns:
    if reg.match(elem):
        toDrop.append(elem)
df_acc_circo = df_acc_circo.drop(columns=toDrop)

#aggregate groups by gravite
acc_circo = df_acc_circo.groupby(['gravite']).sum().sort_values(['2010'], ascending=False)\
            [[col for col in df_acc_circo.columns if IS_DATE.match(col)]]
display(acc_circo)

1992     1993     1994     1995     1996  \
gravite                                                                      
Accidents avec blessés légers  11423.0  11618.0  12649.0  12012.0  11459.0   
Accidents avec blessés graves   4673.0   4261.0   4046.0   4065.0   3695.0   
Accidents mortels                402.0    377.0    377.0    361.0    304.0   

                                  1997     1998     1999     2000     2001  \
gravite                                                                      
Accidents avec blessés légers  11935.0  11848.0  12436.0  12959.0  12809.0   
Accidents avec blessés graves   3634.0   3570.0   3661.0   3554.0   3557.0   
Accidents mortels                302.0    311.0    287.0    316.0    266.0   

                                  2002     2003     2004     2005     2006  \
gravite                                                                      
Accidents avec blessés légers  13064.0  13139.0  12573.0  12153.0  12049.0   
Accidents avec blessés graves   3438.0   3205.0   2995.0   2940.0   2912.0   
Accidents mortels                229.0    272.0    271.0    191.0    192.0   

                                  2007     2008     2009     2010  
gravite                                                            
Accidents avec blessés légers  12303.0  11604.0  11380.0  10605.0  
Accidents avec blessés graves   3068.0   2762.0   2747.0   2460.0  
Accidents mortels                184.0    189.0    191.0    174.0

In [15]:
#test dataset with route type
df_acc_routeType = df_Accident_type_route

#only taking years from 2010 to 2017
toDrop = []
for elem in df_acc_routeType.columns:
    if reg.match(elem):
        toDrop.append(elem)
df_acc_routeType = df_acc_routeType.drop(columns=toDrop)

#aggregate groups by gravite
acc_routeType = df_acc_routeType.groupby(['gravite']).sum().sort_values(['2010'],ascending=False)\
            [[col for col in df_acc_routeType.columns if IS_DATE.match(col)]]
display(acc_routeType)

1992     1993     1994     1995     1996  \
gravite                                                                      
Accidents avec blessés légers  15755.0  15933.0  16967.0  16409.0  15504.0   
Accidents avec blessés graves   6769.0   6258.0   5923.0   5981.0   5512.0   
Accidents mortels                748.0    661.0    637.0    640.0    562.0   

                                  1997     1998     1999     2000     2001  \
gravite                                                                      
Accidents avec blessés légers  16084.0  16207.0  17311.0  17754.0  17942.0   
Accidents avec blessés graves   5438.0   5475.0   5574.0   5428.0   5458.0   
Accidents mortels                553.0    550.0    549.0    555.0    496.0   

                                  2002     2003     2004     2005     2006  \
gravite                                                                      
Accidents avec blessés légers  17866.0  18168.0  17501.0  16719.0  16543.0   
Accidents avec blessés graves   5305.0   5173.0   4912.0   4614.0   4608.0   
Accidents mortels                476.0    499.0    478.0    373.0    340.0   

                                  2007     2008     2009     2010  
gravite                                                            
Accidents avec blessés légers  16790.0  16034.0  15834.0  15214.0  
Accidents avec blessés graves   4760.0   4366.0   4337.0   4082.0  
Accidents mortels                361.0    336.0    335.0    313.0

In [16]:
#test dataset with objects
df_acc_objet = df_Accident_objets

#only taking years from 2010 to 2017
toDrop = []
for elem in df_acc_objet.columns:
    if reg.match(elem):
        toDrop.append(elem)
df_acc_objet = df_acc_objet.drop(columns=toDrop)

#aggregate groups by gravite 
acc_objet = df_acc_objet.groupby(['gravite']).sum().sort_values(['2010'],ascending=False)\
            [[col for col in df_acc_objet.columns if IS_DATE.match(col)]]
display(acc_objet)

1992     1993     1994     1995     1996  \
gravite                                                                      
Accidents avec blessés légers  30149.0  30651.0  33047.0  32067.0  30310.0   
Accidents avec blessés graves  12519.0  11472.0  10803.0  11013.0  10037.0   
Accidents mortels               1321.0   1232.0   1163.0   1135.0   1009.0   

                                  1997     1998     1999     2000     2001  \
gravite                                                                      
Accidents avec blessés légers  31510.0  31686.0  33859.0  34625.0  34814.0   
Accidents avec blessés graves   9966.0   9987.0  10259.0   9868.0   9914.0   
Accidents mortels                972.0   1015.0    996.0   1008.0    902.0   

                                  2002     2003     2004     2005     2006  \
gravite                                                                      
Accidents avec blessés légers  34744.0  35115.0  33672.0  32494.0  32033.0   
Accidents avec blessés graves   9513.0   9156.0   8693.0   8236.0   8264.0   
Accidents mortels                842.0    945.0    860.0    637.0    627.0   

                                  2007     2008     2009     2010  
gravite                                                            
Accidents avec blessés légers  32442.0  30692.0  30474.0  29117.0  
Accidents avec blessés graves   8466.0   7745.0   7685.0   7165.0  
Accidents mortels                627.0    624.0    583.0    564.0

> We know that in 2017, there were 17'799 accidents causing body injuries in Switzerland. Therefore, it means that cantons and routeType datasets are correct! 
However, the dataset of circomstances is lacking of values (missing ~1/2) and the one about objects is containing too much (round 2 times more accidents). **This might be explained by the fact that multiple objects are involved in one accident.**

In [19]:
#aggregate objets dataset in groups by vehicle type
acc_objet = df_acc_objet.groupby(['voiture']).sum().sort_values(['2010'],ascending=False)\
            [[col for col in df_acc_objet.columns if IS_DATE.match(col)]]
display(acc_objet)

1992     1993     1994     1995  \
voiture                                                                  
Voiture de tourisme                 27180.0  26961.0  28208.0  28134.0   
Cycle                                3520.0   3593.0   3900.0   3637.0   
Piéton                               3157.0   3052.0   2993.0   3165.0   
Motocycle de plus de 125 cm3         2068.0   2066.0   2071.0   1950.0   
Véhicule de transport de choses      2440.0   2391.0   2408.0   2516.0   
Motocycle jusqu'à 125 cm3            1999.0   1887.0   1855.0   1635.0   
Motocycle léger                       270.0    276.0    272.0    259.0   
Cyclomoteur                          2239.0   2001.0   2162.0   1867.0   
Autre et inconnu                      552.0    536.0    611.0    483.0   
Véhicule de transport de personnes    461.0    474.0    441.0    464.0   
Autre véhicule non motorisé           103.0    118.0     92.0    105.0   

                                       1996     1997     1998     1999  \
voiture                                                                  
Voiture de tourisme                 26318.0  26636.0  27603.0  29248.0   
Cycle                                3586.0   3869.0   3510.0   3546.0   
Piéton                               2913.0   2888.0   2905.0   3043.0   
Motocycle de plus de 125 cm3         1686.0   1896.0   1640.0   1768.0   
Véhicule de transport de choses      2271.0   2270.0   2314.0   2588.0   
Motocycle jusqu'à 125 cm3            1567.0   1748.0   1620.0   1582.0   
Motocycle léger                       267.0    361.0    560.0    802.0   
Cyclomoteur                          1740.0   1721.0   1546.0   1475.0   
Autre et inconnu                      502.0    528.0    459.0    528.0   
Véhicule de transport de personnes    410.0    420.0    407.0    416.0   
Autre véhicule non motorisé            96.0    111.0    124.0    118.0   

                                       2000     2001     2002     2003  \
voiture                                                                  
Voiture de tourisme                 29617.0  29919.0  29466.0  29087.0   
Cycle                                3503.0   3358.0   3394.0   3579.0   
Piéton                               3003.0   2808.0   2769.0   2645.0   
Motocycle de plus de 125 cm3         1801.0   1856.0   1832.0   2302.0   
Véhicule de transport de choses      2545.0   2473.0   2411.0   2406.0   
Motocycle jusqu'à 125 cm3            1621.0   1694.0   1605.0   1664.0   
Motocycle léger                      1088.0   1266.0   1505.0   1430.0   
Cyclomoteur                          1347.0   1225.0   1165.0   1038.0   
Autre et inconnu                      479.0    529.0    461.0    486.0   
Véhicule de transport de personnes    387.0    356.0    341.0    363.0   
Autre véhicule non motorisé           110.0    146.0    150.0    216.0   

                                       2004     2005     2006     2007  \
voiture                                                                  
Voiture de tourisme                 27639.0  26246.0  25701.0  25958.0   
Cycle                                3390.0   3373.0   3417.0   3513.0   
Piéton                               2634.0   2514.0   2653.0   2679.0   
Motocycle de plus de 125 cm3         2268.0   2290.0   2239.0   2422.0   
Véhicule de transport de choses      2279.0   2285.0   2403.0   2313.0   
Motocycle jusqu'à 125 cm3            1921.0   1857.0   1985.0   2014.0   
Motocycle léger                      1097.0    888.0    746.0    742.0   
Cyclomoteur                           962.0    858.0    781.0    851.0   
Autre et inconnu                      453.0    497.0    456.0    540.0   
Véhicule de transport de personnes    392.0    383.0    372.0    367.0   
Autre véhicule non motorisé           190.0    176.0    171.0    136.0   

                                       2008     2009     2010  
voiture                                                        
Voiture de tourisme                 24392.0  23912.0  22889

> In previous table, we can see that in 2017, more cars were involved in accidents than the total number of accidents (18981>17799). This illustrates that many vehicles are implied in accidents. 

In [22]:
totCrash=sum(acc_cantons['2010'])
totObjects = sum(acc_objet['2010'])
objetPerCrash = totObjects/totCrash

print('In average, a crash was involving %.2f objects in 2017.' % objetPerCrash)

In average, a crash was involving 1.88 objects in 2017.


Now let's study the case of circonstances dataset:

In [23]:
#aggregate groups by circonstances
acc_circo = df_acc_circo.groupby(['circonstance']).sum().sort_values(['2010'], ascending=False)\
            [[col for col in df_acc_circo.columns if IS_DATE.match(col)]]
display(acc_circo)

1992     1993     1994     1995     1996  \
circonstance                                                                  
Autre facteur d'influence       15441.0  15291.0  16094.0  15570.0  14649.0   
Alcool                            980.0    894.0    896.0    787.0    731.0   
Drogues                            40.0     40.0     49.0     42.0     33.0   
Alcool + droques                   15.0      9.0     12.0     18.0     27.0   
Médicaments                        10.0     11.0     12.0     10.0      8.0   
Alcool + médicaments                5.0      7.0      4.0      6.0      4.0   
Drogues et médicaments              7.0      3.0      5.0      4.0      5.0   
Alcool + drogues + médicaments      0.0      1.0      0.0      1.0      1.0   

                                   1997     1998     1999     2000     2001  \
circonstance                                                                  
Autre facteur d'influence       15051.0  14853.0  15445.0  15954.0  15776.0   
Alcool                            735.0    798.0    861.0    787.0    763.0   
Drogues                            46.0     40.0     33.0     46.0     41.0   
Alcool + droques                   18.0     16.0     21.0     15.0     26.0   
Médicaments                        14.0     11.0     13.0     17.0     15.0   
Alcool + médicaments                3.0      6.0      7.0      8.0      6.0   
Drogues et médicaments              2.0      5.0      3.0      2.0      5.0   
Alcool + drogues + médicaments      2.0      0.0      1.0      0.0      0.0   

                                   2002     2003     2004     2005     2006  \
circonstance                                                                  
Autre facteur d'influence       15852.0  15813.0  15058.0  14469.0  14219.0   
Alcool                            778.0    705.0    654.0    672.0    805.0   
Drogues                            50.0     50.0     57.0     67.0     63.0   
Alcool + droques                   24.0     24.0     27.0     39.0     35.0   
Médicaments                        15.0     13.0     30.0     23.0     17.0   
Alcool + médicaments                7.0      7.0      8.0      7.0     11.0   
Drogues et médicaments              4.0      4.0      5.0      6.0      2.0   
Alcool + drogues + médicaments      1.0      0.0      0.0      1.0      1.0   

                                   2007     2008     2009     2010  
circonstance                                                        
Autre facteur d'influence       14711.0  13709.0  13528.0  12446.0  
Alcool                            720.0    737.0    661.0    658.0  
Drogues                            57.0     52.0     59.0     63.0  
Alcool + droques                   27.0     17.0     18.0     28.0  
Médicaments                        21.0     23.0     32.0     23.0  
Alcool + médicaments               10.0     14.0     19.0     17.0  
Drogues et médicaments              6.0      2.0      0.0      3.0  
Alcool + drogues + médicaments      3.0      1.0      1.0      1.0

In [329]:
totCirco = sum(acc_circo['2017'])
totCirco

11754.0

> Hmmm... it looks like this dataset is incomplete! It provides only 11754 values out of 17799 accidents... Another possibility would be that "Autres facteurs d'influence" doesn't contain "No influence factor" for example. 

In [330]:
#aggregate circo dataset in groups by context
acc_circo = df_acc_circo.groupby(['contexte']).sum().sort_values(['2010'], ascending=False)\
            [[col for col in df_acc_circo.columns if IS_DATE.match(col)]]
display(acc_circo)

2010    2011    2012  \
contexte                                                                 
Tamponnement                                    4061.0  4062.0  3822.0   
Changement de direction (en obliquant)          3887.0  3855.0  3457.0   
Accident concernant des piétons                 2251.0  2187.0  2113.0   
À une bifurcation sans changement de direction  1058.0  1027.0   979.0   
Lors de croisement (longitudinal)                833.0   805.0   762.0   
Autre                                            602.0   546.0   522.0   
Dépassement                                      465.0   497.0   415.0   
Avec un animal                                    82.0    74.0    91.0   

                                                  2013    2014    2015  \
contexte                                                                 
Tamponnement                                    3659.0  3882.0  3701.0   
Changement de direction (en obliquant)          3383.0  3535.0  3481.0   
Accident concernant des piétons                 2100.0  2081.0  2110.0   
À une bifurcation sans changement de direction   922.0   916.0   910.0   
Lors de croisement (longitudinal)                675.0   649.0   665.0   
Autre                                            489.0   497.0   537.0   
Dépassement                                      434.0   416.0   435.0   
Avec un animal                                    72.0    89.0    99.0   

                                                  2016    2017  
contexte                                                        
Tamponnement                                    3862.0  3651.0  
Changement de direction (en obliquant)          3438.0  3464.0  
Accident concernant des piétons                 2073.0  2097.0  
À une bifurcation sans changement de direction   881.0   865.0  
Lors de croisement (longitudinal)                655.0   703.0  
Autre                                            530.0   479.0  
Dépassement                                      402.0   417.0  
Avec un animal                                    74.0    78.0

> Still the same issue, and 'Autre' is even included... I think there miss values

## B. Analyse on the actor of car accident

> Maybe our previous results are false since it's relative to the population but we don't know the part of the population who has a driver licence. So now let study the pur data. We prepare the dataset to a pourcentage relative to the number of car accident during the year.

In [331]:
df_Accident = df_Accident_objets[(df_Accident_objets.faute_index==1) & (df_Accident_objets.voiture_index==210)]
for column in df_Accident.columns:
    if IS_DATE.match(column):
        df_Accident[column] = df_Accident[column].apply(lambda x: x*100/(df_Accident[column].sum()))

C:\Users\Clementg20\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.


### B.1. Sexe and Age of drivers

In [332]:
display(df_Accident.groupby(['sexe']).sum().sort_values(['2017'], ascending=False)\
        [[col for col in df_Accident.columns if IS_DATE.match(col)]])

1992       1993       1994       1995       1996       1997  \
sexe                                                                        
Homme    72.550478  71.591507  70.455072  70.507938  70.061164  69.991251   
Femme    25.716410  26.764930  27.860120  27.880954  28.211110  28.215223   
Inconnu   1.733112   1.643563   1.684808   1.611108   1.727726   1.793526   

              1998       1999       2000       2001    ...           2008  \
sexe                                                   ...                  
Homme    69.749787  68.984757  68.619270  68.311121    ...      63.908187   
Femme    28.719786  29.209088  29.723178  30.081756    ...      34.028062   
Inconnu   1.530426   1.806155   1.657552   1.607123    ...       2.063752   

              2009       2010       2011       2012       2013       2014  \
sexe                                                                        
Homme    64.488518  63.897465  63.744292  63.759194  63.543805  62.981812   
Femme    33.778706  33.991527  34.056317  34.282059  34.641596  35.040881   
Inconnu   1.732777   2.111007   2.199391   1.958746   1.814599   1.977307   

              2015       2016       2017  
sexe                                      
Homme    61.982833  62.231615  61.811600  
Femme    35.785408  35.807270  36.309575  
Inconnu   2.231760   1.961116   1.878824  

[3 rows x 26 columns]

In [333]:
display(df_Accident.groupby(['age']).sum().sort_values(['2017'], ascending=False)\
        [[col for col in df_Accident.columns if IS_DATE.match(col)]].head(5))

1992       1993       1994       1995       1996  \
age                                                                     
21-29 ans       32.362867  30.806574  29.845994  29.105396  27.340942   
30-39 ans       19.567580  19.933322  21.120889  21.020564  21.596570   
40-49 ans       13.933535  14.704334  14.578509  14.880778  15.070307   
50-59 ans        9.655094  10.352693  10.398333  10.914523  10.946466   
70 ans et plus   5.062060   5.457098   5.679713   5.958170   6.343401   

                     1997       1998       1999       2000       2001  \
age                                                                     
21-29 ans       26.390451  26.497024  26.304285  24.575746  24.599619   
30-39 ans       21.853518  21.559577  22.530917  21.790607  21.581364   
40-49 ans       14.554431  15.547188  14.713834  15.295710  15.785642   
50-59 ans       11.604799  11.387101  11.285591  11.805830  11.132266   
70 ans et plus   6.749156   6.170290   6.016681   6.833174   6.921268   

                  ...           2008       2009       2010       2011  \
age               ...                                                   
21-29 ans         ...      23.934069  23.458594  22.574855  22.549467   
30-39 ans         ...      17.872224  17.418232  17.175271  16.651446   
40-49 ans         ...      17.647459  17.390397  17.893301  16.849315   
50-59 ans         ...      12.409753  12.588727  12.938896  12.815830   
70 ans et plus    ...       7.982564   8.608212   8.738422   9.634703   

                     2012       2013       2014       2015       2016  \
age                                                                     
21-29 ans       21.410297  21.873717  21.341565  20.849785  20.481826   
30-39 ans       17.276943  17.415223  16.827966  16.334764  17.100592   
40-49 ans       17.628718  16.725511  16.385783  16.377682  15.198648   
50-59 ans       13.383435  13.334428  14.300017  14.909871  14.784446   
70 ans et plus   9.881676  10.181460  10.737527  11.527897  12.375317   

                     2017  
age                        
21-29 ans       20.744635  
30-39 ans       17.452383  
40-49 ans       14.935792  
50-59 ans       14.909937  
70 ans et plus  12.548479  

[5 rows x 26 columns]

In [335]:
display(df_Accident.groupby(['sexe','age']).sum().sort_values(['2017'], ascending=False)\
        [[col for col in df_Accident.columns if IS_DATE.match(col)]].head(5))

1992       1993       1994       1995       1996  \
sexe  age                                                                
Homme 21-29 ans  23.560030  22.272913  21.271422  20.692483  19.307649   
      30-39 ans  14.316765  14.218869  14.404817  14.623001  15.246863   
      40-49 ans   9.775210   9.972510   9.981473  10.176343  10.259159   
      50-59 ans   7.058285   7.498392   7.596109   7.686449   7.667570   
Femme 21-29 ans   8.802837   8.533661   8.574572   8.412912   8.033293   

                      1997       1998       1999       2000       2001  \
sexe  age                                                                
Homme 21-29 ans  18.416448  18.899551  18.349152  17.139313  16.933587   
      30-39 ans  15.223097  14.696951  15.478861  14.083554  14.497704   
      40-49 ans   9.980002  10.737277   9.882082  10.385071  10.549894   
      50-59 ans   8.248969   7.864691   7.794075   8.276484   7.576436   
Femme 21-29 ans   7.974003   7.597474   7.955134   7.436432   7.666032   

                   ...           2008       2009       2010       2011  \
sexe  age          ...                                                   
Homme 21-29 ans    ...      15.175044  15.414057  14.669347  14.436834   
      30-39 ans    ...      10.999864  11.113431  10.863790  10.471842   
      40-49 ans    ...      11.176951  10.946416  11.208444  10.304414   
      50-59 ans    ...       8.234573   8.030619   8.573275   8.287671   
Femme 21-29 ans    ...       8.759025   8.044537   7.905507   8.112633   

                      2012       2013       2014       2015       2016  \
sexe  age                                                                
Homme 21-29 ans  13.751199  13.810658  13.340564  12.918455  12.907861   
      30-39 ans  10.833067  11.060021  10.946104  10.017167  10.431107   
      40-49 ans  11.096898  10.534527   9.944936   9.682403   9.315300   
      50-59 ans   8.786377   8.522867   9.227432   9.665236   9.323753   
Femme 21-29 ans   7.659098   8.063059   8.001001   7.931330   7.573964   

                      2017  
sexe  age                   
Homme 21-29 ans  12.436439  
      30-39 ans  11.247091  
      40-49 ans   9.221753  
      50-59 ans   9.109713  
Femme 21-29 ans   8.308196  

[5 rows x 26 columns]

### B.2. Licence term of drivers

In [337]:
mapping={}
index_permis = df_Accident_objets['annee_permis_index'].unique()
for i in range(len(index_permis)):
    if ((index_permis[i]==999) | (index_permis[i] == 16)):
        mapping[index_permis[i]] = 1
    else:
        mapping[index_permis[i]] = index_permis[i+1] - index_permis[i]


df_permis = df_Accident.copy()

df_permis = df_Accident[(df_Accident.faute_index==1)\
                            & (df_Accident.voiture_index==210)]
df_permis = df_permis.groupby(['annee_permis','annee_permis_index']).sum().sort_values(['2017'], ascending=False)
dates = [col for col in df_permis.columns if IS_DATE.match(col)]

for i in range(len(df_permis.index)):
    df_permis.loc[df_permis.index.values[i]] = df_permis.loc[df_permis.index.values[i]] / mapping[df_permis.index.values[i][1]]
df_permis = df_permis[[col for col in df_permis.columns if IS_DATE.match(col)]].sort_values(['2017'], ascending=False)
display(df_permis)

1992       1993       1994       1995  \
annee_permis   annee_permis_index                                               
16 ans et plus 16.0                34.101699  36.129145  36.903659  38.180327   
0 an           0.0                  8.940113   7.527636   7.399259   7.375945   
1 an           1.0                  7.029686   6.720477   5.969199   5.794130   
2 ans          2.0                  5.811360   5.878224   5.743400   5.184838   
3 ans          3.0                  5.685523   5.100310   5.123900   4.845041   
Inconnu        999.0                3.008637   2.702228   2.703798   2.659793   
4 ans          4.0                  4.690271   4.749371   4.753358   4.405648   
5 ans          5.0                  4.055368   4.000702   4.047013   3.972113   
6-10 ans       6.0                  3.187096   3.199392   3.105604   3.191751   
11-15 ans      11.0                 2.148373   2.238989   2.365679   2.324682   

                                        1996       1997       1998       1999  \
annee_permis   annee_permis_index                                               
16 ans et plus 16.0                39.239549  40.826147  40.428762  40.103538   
0 an           0.0                  8.506211   7.892763   8.368760   8.231234   
1 an           1.0                  5.511066   5.593051   6.145998   6.045441   
2 ans          2.0                  4.281481   4.468191   5.198591   5.050331   
3 ans          3.0                  4.022952   3.918260   3.977894   4.377337   
Inconnu        999.0                2.818589   2.730909   2.459614   2.876043   
4 ans          4.0                  4.098619   3.993251   3.516337   3.606557   
5 ans          5.0                  4.054480   3.712036   3.273412   3.111878   
6-10 ans       6.0                  3.097295   2.977128   2.991619   2.958873   
11-15 ans      11.0                 2.396116   2.395951   2.334507   2.360656   

                                        2000       2001    ...           2008  \
annee_permis   annee_permis_index                          ...                  
16 ans et plus 16.0                42.769352  42.826744    ...      47.370930   
0 an           0.0                  8.597846   8.746780    ...       7.553467   
1 an           1.0                  5.677398   6.081308    ...       5.373927   
2 ans          2.0                  4.617466   4.810169    ...       4.502111   
3 ans          3.0                  3.957828   3.936611    ...       3.889116   
Inconnu        999.0                2.892259   2.743868    ...       3.698406   
4 ans          4.0                  3.574449   3.572629    ...       3.555374   
5 ans          5.0                  2.914811   3.018255    ...       3.119466   
6-10 ans       6.0                  2.746800   2.571397    ...       2.456069   
11-15 ans      11.0                 2.252918   2.281330    ...       1.731372   

                                        2009       2010       2011       2012  \
annee_permis   annee_permis_index                                               
16 ans et plus 16.0                48.517745  49.019889  49.398782  50.623601   
0 an           0.0                  7.787056   7.331084   7.480974   7.331308   
1 an           1.0                  5.170494   4.724636   5.167428   5.132715   
2 ans          2.0                  4.370216   4.143032   3.873668   3.837544   
3 ans          3.0                  4.050104   3.324478   3.485540   3.445795   
Inconnu        999.0                3.215031   3.740935   4.041096   3.725616   
4 ans          4.0                  3.208072   3.288576   3.302892   2.774224   
5 ans          5.0                  2.901879   3.123429   2.777778   2.518388   
6-10 ans       6.0                  2.336813   2.487255   2.316591   2.251359   
11-15 ans      11.0                 1.819068   1.773533   1.777778   1.870803   

                                        2013       2014       2015       2016  \
annee_permis   annee_permis_index                                  

## C. Analyse on the actors of accidents relative to the population

### C.0. Normalize the data by the population

#### C0.1. Get the total swiss population

In [341]:
ages = build_age(df_Accident_objets.age_index.unique())
for age in ages:
    df_Population_2010[str(age[0])] = df_Population_2010.loc[:,age[1:]].sum(axis=1)

toKeep = []
for age in ages :
    toKeep.append(str(age[0]))
toKeep.append('annee')
toKeep.append('sexe')

df_pop_Suisse = df_Population_2010[(df_Population_2010['canton'] == 'Suisse') & (df_Population_2010['sexe'] != 'Sexe - Total')\
                                     & (df_Population_2010.etat_civil == "Etat civil - Total")]

df_pop_Suisse = df_pop_Suisse[toKeep]
df_pop_Suisse['annee'] = df_pop_Suisse['annee'].map(int).map(str)
df_pop_Suisse = df_pop_Suisse.groupby(['annee','sexe']).sum()
df_pop_Suisse = df_pop_Suisse.transpose()
df_pop_Suisse.index = df_pop_Suisse.index.map(float)

In [342]:
df_Suisse_1992 = df_Population_age_1992[(df_Population_age_1992.sexe != 'Sexe - Total') & (df_Population_age_1992.age != 'Age - Total')]
df_Suisse_1992 = df_Suisse_1992.groupby(['age','sexe']).sum().transpose()

for age in ages:
    for s in df_Population_age_1992[df_Population_age_1992.sexe != 'Sexe - Total'].sexe.unique():
        index_to = [(annee,s) for annee in age[1:]]
        df_Suisse_1992[(str(age[0]),s)] = df_Suisse_1992.loc[:,index_to].sum(axis=1)
df_Suisse_1992 = df_Suisse_1992.transpose().reset_index().set_index('age').transpose()

toKeep = []
for age in ages :
    toKeep.append(str(age[0]))
df_Suisse_1992 = df_Suisse_1992[toKeep]
df_Suisse_1992 = df_Suisse_1992.transpose()

for year in range(1992,2010):
    for sexe in np.unique([sexe[1] for sexe in df_pop_Suisse.columns.values]):
        df_pop_Suisse[(str(year),sexe)] = df_Suisse_1992.loc[df_Suisse_1992.sexe==sexe][str(year)]
display(df_pop_Suisse.head(5))

C:\Users\Clementg20\Anaconda3\lib\site-packages\pandas\core\indexing.py:979: FutureWarning: 
Passing list-likes to .loc or [] with any missing label will raise
KeyError in the future, you can use .reindex() as an alternative.

See the documentation here:
https://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike
  return self._getitem_nested_tuple(tup)


annee      2010                2011                2012                2013  \
sexe      Femme     Homme     Femme     Homme     Femme     Homme     Femme   
0.0    191197.0  202097.0  195313.0  206364.0  198173.0  210120.0  201280.0   
5.0     37402.0   39653.0   37807.0   40014.0   38743.0   40693.0   39450.0   
6.0    149641.0  157585.0  150877.0  158790.0  152374.0  160288.0  154842.0   
10.0   203641.0  214404.0  200464.0  211509.0  198210.0  209411.0  196714.0   
15.0   128790.0  135406.0  128497.0  135222.0  127946.0  134486.0  127218.0   

annee                2014             ...      2005            2006          \
sexe      Homme     Femme     Homme   ...     Femme   Homme   Femme   Homme   
0.0    213036.0  204828.0  216474.0   ...    177363  187819  177587  187995   
5.0     42127.0   40359.0   42763.0   ...     36886   39202   36023   38385   
6.0    162938.0  157261.0  166440.0   ...    157083  166095  154375  163502   
10.0   207509.0  196515.0  207080.0   ...    211676  223411  209411  221047   
15.0   134069.0  125857.0  132988.0   ...    129832  137137  132068  139716   

annee    2007            2008            2009          
sexe    Femme   Homme   Femme   Homme   Femme   Homme  
0.0    178473  188725  181376  191869  184239  195325  
5.0     35991   38393   36173   38361   36626   38626  
6.0    151269  160537  149927  159496  148587  158177  
10.0   206647  218410  205007  217162  203635  215747  
15.0   133365  141070  133177  140852  131770  139313  

[5 rows x 52 columns]

> `df_pop_Suisse` contains the Swiss population by **year** and **sexe** on *columns* and by **age** range on *rows*

#### C.0.1. Normalize

In [343]:
df_Accident_victimes_normalized = df_Accident_victimes.copy()
df_Accident_victimes_normalized = df_Accident_victimes_normalized[df_Accident_victimes_normalized['sexe_index']!=9]
for year,sexe in df_pop_Suisse.columns.values:
    for age in df_pop_Suisse.index.values:
        df_Accident_victimes_normalized.loc[(df_Accident_victimes_normalized.age_index==age) &\
                                           (df_Accident_victimes_normalized.sexe==sexe),year]\
        = df_Accident_victimes_normalized[year]*10000 / df_pop_Suisse.loc[df_pop_Suisse.index==age,year][sexe].values[0]

df_Accident_victimes_normalized = df_Accident_victimes_normalized[df_Accident_victimes_normalized.age_index != 999]
display(df_Accident_victimes_normalized.head(2))

voiture_index              voiture  sexe_index   sexe  role_index  \
0          210.0  Voiture de tourisme         1.0  Homme       300.0   
1          210.0  Voiture de tourisme         1.0  Homme       300.0   

         role  age_index            age  consequence_index   consequence  \
0  Conducteur        0.0  jusqu'à 4 ans              315.0  Blessé léger   
1  Conducteur        0.0  jusqu'à 4 ans              316.0  Blessé grave   

   ...   2008  2009  2010  2011      2012  2013  2014  2015  2016  2017  
0  ...    0.0   0.0   0.0   0.0  0.047592   0.0   0.0   0.0   0.0   0.0  
1  ...    0.0   0.0   0.0   0.0  0.000000   0.0   0.0   0.0   0.0   0.0  

[2 rows x 36 columns]

> `df_Accident_acteurs_normalized` contains the proportion in term of **age range** and **sexe** of the population involved in accident

In [344]:
df_Accident_objets_normalized = df_Accident_objets.copy()
df_Accident_objets_normalized = df_Accident_objets_normalized[df_Accident_objets_normalized['sexe_index']!=9]
for year,sexe in df_pop_Suisse.columns.values:
    for age in df_pop_Suisse.index.values:
        df_Accident_objets_normalized.loc[(df_Accident_objets_normalized.age_index==age) &\
                                           (df_Accident_objets_normalized.sexe==sexe),year]\
        = df_Accident_objets_normalized[year]*10000 / df_pop_Suisse.loc[df_pop_Suisse.index==age,year][sexe].values[0]

df_Accident_objets_normalized = df_Accident_objets_normalized[df_Accident_objets_normalized.age_index != 999]
display(df_Accident_objets_normalized.head(2))

voiture_index              voiture  age_index            age  sexe_index  \
0          210.0  Voiture de tourisme        0.0  Jusqu'à 4 ans         1.0   
1          210.0  Voiture de tourisme        0.0  Jusqu'à 4 ans         1.0   

    sexe  annee_permis_index annee_permis  faute_index  \
0  Homme                 0.0         0 an          0.0   
1  Homme                 0.0         0 an          0.0   

                           faute  ...   2008 2009  2010  2011  2012  2013  \
0  Objet sans faute ni influence  ...    0.0  0.0   0.0   0.0   0.0   0.0   
1  Objet sans faute ni influence  ...    0.0  0.0   0.0   0.0   0.0   0.0   

   2014  2015  2016  2017  
0   0.0   0.0   0.0   0.0  
1   0.0   0.0   0.0   0.0  

[2 rows x 38 columns]

> `df_Accident_objets_normalized` contains the proportion in term of **age range** and **sexe** of the population involved in accident

### C.1. Sexe and Age of drivers

#### C.1.1. Sexe and Age together

In [346]:
df_conducteur = df_Accident_objets_normalized[(df_Accident_objets_normalized.faute_index==1)\
                                           & (df_Accident_objets_normalized.voiture_index==210)]
df_conducteur = df_conducteur.groupby(['sexe','age']).sum().sort_values(['2017'], ascending=False)[[col for col in df_conducteur.columns if IS_DATE.match(col)]]
display(df_conducteur.head(5))

1992        1993        1994        1995       1996  \
sexe  age                                                                    
Homme 20 ans     127.971524  117.852950  125.859697  115.426207  92.977202   
      18-19 ans   91.262445   85.548212   79.800978   79.263615  84.056135   
      21-29 ans   79.562532   75.568450   75.104204   74.622609  67.304099   
Femme 20 ans      41.204299   47.217153   39.882947   42.637725  36.786490   
Homme 30-39 ans   43.852580   41.868027   41.974276   41.392829  39.754567   

                       1997        1998        1999        2000        2001  \
sexe  age                                                                     
Homme 20 ans     111.426699  117.792183  130.883766  118.059614  131.882700   
      18-19 ans   78.972986   86.307553   87.905889   97.575091   99.309354   
      21-29 ans   67.516639   73.536188   76.999389   74.377638   72.578555   
Femme 20 ans      33.910984   38.310314   41.470970   47.580900   48.580501   
Homme 30-39 ans   40.106226   39.883317   44.419721   41.375975   42.328752   

                   ...           2008       2009       2010       2011  \
sexe  age          ...                                                   
Homme 20 ans       ...      76.846329  68.903970  60.461671  61.102625   
      18-19 ans    ...      68.300948  65.115308  55.686596  59.962381   
      21-29 ans    ...      50.191485  48.996512  42.968643  39.242457   
Femme 20 ans       ...      33.497884  37.035442  32.691987  24.248202   
Homme 30-39 ans    ...      28.826623  28.567289  27.008356  24.289025   

                      2012       2013       2014       2015       2016  \
sexe  age                                                                
Homme 20 ans     55.482377  45.599542  47.268263  40.245925  43.554182   
      18-19 ans  49.946674  47.051076  43.540140  37.905391  37.816045   
      21-29 ans  35.122254  33.921480  31.927447  29.644892  30.092583   
Femme 20 ans     27.287386  26.002259  31.368168  26.294974  25.347733   
Homme 30-39 ans  23.592174  23.070797  22.080219  19.292668  20.161949   

                      2017  
sexe  age                   
Homme 20 ans     39.736442  
      18-19 ans  35.445563  
      21-29 ans  28.468726  
Femme 20 ans     25.532472  
Homme 30-39 ans  21.087092  

[5 rows x 26 columns]

#### C.1.2. Sexe only

In [347]:
display(df_conducteur.groupby(['sexe']).sum().sort_values(['2017'], ascending=False))

1992        1993        1994        1995        1996        1997  \
sexe                                                                            
Homme  472.150909  452.039610  453.100949  442.028498  405.744536  421.970174   
Femme  155.807492  159.759884  158.269271  158.433251  150.791784  144.482553   

             1998        1999        2000        2001     ...            2008  \
sexe                                                      ...                   
Homme  439.099231  462.567421  464.760519  472.880121     ...      318.479298   
Femme  158.113546  169.336094  176.822786  182.652496     ...      145.942061   

             2009        2010        2011        2012        2013        2014  \
sexe                                                                            
Homme  304.020417  275.120250  267.219327  244.140696  226.372270  219.320518   
Femme  146.576781  137.243465  121.364819  121.848155  119.164725  119.277469   

             2015        2016        2017  
sexe                                       
Homme  200.363493  205.480487  195.489628  
Femme  112.998868  114.132016  114.017821  

[2 rows x 26 columns]

#### C.1.3. Age only

In [348]:
display(df_conducteur.groupby(['age']).sum().sort_values(['2017'], ascending=False))

1992        1993        1994        1995        1996  \
age                                                                          
20 ans          169.175823  165.070103  165.742644  158.063932  129.763693   
18-19 ans       118.734644  110.401491  105.036700  101.916682  110.876743   
21-29 ans       110.318358  105.070017  105.673692  105.005076   95.110129   
30-39 ans        60.870922   59.570028   62.495261   60.304638   56.918102   
40-49 ans        47.902619   49.357171   49.300135   49.566926   46.548065   
70 ans et plus   30.850933   32.166982   32.770495   33.746316   32.462610   
50-59 ans        43.533557   44.751647   44.326456   44.851796   40.830010   
60-69 ans        37.451596   37.405979   38.351241   38.253963   36.311814   
15-17 ans         5.888159    5.351004    4.688106    6.024098    5.106690   
10-14 ans         3.169508    2.655073    2.939557    2.728322    2.515621   
5 ans             0.000000    0.000000    0.000000    0.000000    0.000000   
6-9 ans           0.062282    0.000000    0.000000    0.000000    0.000000   
Jusqu'à 4 ans     0.000000    0.000000    0.045932    0.000000    0.092845   

                      1997        1998        1999        2000        2001  \
age                                                                          
20 ans          145.337682  156.102496  172.354736  165.640514  180.463201   
18-19 ans        99.097579  112.813000  115.999601  123.647997  129.953734   
21-29 ans        96.397309  102.762975  110.150468  106.550017  105.702915   
30-39 ans        58.022168   58.840376   64.887338   64.165233   63.343158   
40-49 ans        45.277973   49.482859   48.971764   51.216346   51.918942   
70 ans et plus   34.263381   31.303280   32.010362   36.557740   35.770656   
50-59 ans        42.650705   42.042554   43.018270   45.018277   42.012526   
60-69 ans        38.419658   36.383173   37.071092   38.978375   36.780571   
15-17 ans         4.185377    4.802986    4.434801    5.908898    6.225329   
10-14 ans         2.753784    2.628577    3.005084    3.661072    3.361585   
5 ans             0.000000    0.000000    0.000000    0.238834    0.000000   
6-9 ans           0.000000    0.000000    0.000000    0.000000    0.000000   
Jusqu'à 4 ans     0.047111    0.050500    0.000000    0.000000    0.000000   

                  ...            2008        2009       2010       2011  \
age               ...                                                     
20 ans            ...      110.344213  105.939412  93.153658  85.350827   
18-19 ans         ...       88.450343   87.723851  76.853005  78.306997   
21-29 ans         ...       79.556707   74.902028  66.802340  61.984712   
30-39 ans         ...       47.015815   44.953023  42.970642  38.941965   
40-49 ans         ...       40.864717   38.876569  38.355462  33.981593   
70 ans et plus    ...       29.158526   30.126845  28.681755  29.111991   
50-59 ans         ...       36.034670   35.148150  33.960527  30.868903   
60-69 ans         ...       30.772544   31.221792  30.229972  28.658326   
15-17 ans         ...        1.933272    1.659177   1.263071   1.187110   
10-14 ans         ...        0.238434    0.046351   0.093282   0.191722   
5 ans             ...        0.000000    0.000000   0.000000   0.000000   
6-9 ans           ...        0.000000    0.000000   0.000000   0.000000   
Jusqu'à 4 ans     ...        0.052119    0.000000   0.000000   0.000000   

                     2012       2013       2014       2015       2016  \
age                                                                     
20 ans          82.769763  71.601801  78.636430  66.540899  68.901914   
18-19 ans       71.356608  68.629788  60.799182  54.831518  56.345653   
21-29 ans       55.321889  54.301182  51.660851  48.558544  48.465890   
30-39 ans       37.982282  36.682755  34.304946  31.851688  33.454870   
40-49 ans       33.931181  31.569275  30.783862  30.298006  28.926313   
70 ans et plus  27.407709  26.992079  26.63

> Number of **conductors** who drive a **normal car** who are **responsible** an accident with victimes. Numbers are 1/10,000 citizens of that categorie of citizen

### C.2. Licence term of drivers

In [349]:
mapping={}
index_permis = df_Accident_objets['annee_permis_index'].unique()
for i in range(len(index_permis)):
    if ((index_permis[i]==999) | (index_permis[i] == 16)):
        mapping[index_permis[i]] = 1
    else:
        mapping[index_permis[i]] = index_permis[i+1] - index_permis[i]


df_permis = df_Accident_objets_normalized.copy()

df_permis = df_Accident_objets_normalized[(df_Accident_objets_normalized.faute_index==1)\
                            & (df_Accident_objets_normalized.voiture_index==210)]
df_permis = df_permis.groupby(['annee_permis','annee_permis_index']).sum().sort_values(['2017'], ascending=False)
dates = [col for col in df_permis.columns if IS_DATE.match(col)]

for i in range(len(df_permis.index)):
    df_permis.loc[df_permis.index.values[i]] = df_permis.loc[df_permis.index.values[i]] / mapping[df_permis.index.values[i][1]]
df_permis = df_permis[[col for col in df_permis.columns if IS_DATE.match(col)]].sort_values(['2017'], ascending=False)
display(df_permis)

1992        1993        1994  \
annee_permis   annee_permis_index                                       
16 ans et plus 16.0                151.830988  155.492622  157.363385   
0 an           0.0                 142.498262  123.374769  123.556499   
1 an           1.0                 115.056016  116.987970  105.379789   
2 ans          2.0                  60.729267   61.781925   67.446199   
3 ans          3.0                  21.661587   19.040298   19.630905   
4 ans          4.0                  16.808797   16.995324   17.814826   
5 ans          5.0                  14.839632   14.184733   15.103948   
6-10 ans       6.0                  11.842954   11.878440   11.847788   
11-15 ans      11.0                  7.140738    7.221523    7.627887   
Inconnu        999.0                 9.615391    8.442038    7.696294   

                                         1995        1996        1997  \
annee_permis   annee_permis_index                                       
16 ans et plus 16.0                159.403876  150.194756  156.738364   
0 an           0.0                 128.246876  134.726054  128.560318   
1 an           1.0                 103.902797   91.447105   92.487221   
2 ans          2.0                  53.471419   37.813390   45.859647   
3 ans          3.0                  18.491277   15.035516   15.244564   
4 ans          4.0                  16.588059   14.531135   15.446187   
5 ans          5.0                  15.069014   14.938232   13.351162   
6-10 ans       6.0                  12.109018   11.073046   11.179562   
11-15 ans      11.0                  7.237492    6.932951    6.924922   
Inconnu        999.0                 8.555882    7.820147    8.242843   

                                         1998        1999        2000  \
annee_permis   annee_permis_index                                       
16 ans et plus 16.0                155.433811  160.281814  173.458723   
0 an           0.0                 138.590490  148.358898  156.150508   
1 an           1.0                 105.406004  114.483530  107.449539   
2 ans          2.0                  53.980696   55.626946   53.101017   
3 ans          3.0                  15.719121   18.181270   17.416233   
4 ans          4.0                  13.960276   15.545337   15.618946   
5 ans          5.0                  12.610990   13.472996   12.903029   
6-10 ans       6.0                  11.864094   12.247334   11.829183   
11-15 ans      11.0                  7.010503    7.381686    7.136687   
Inconnu        999.0                 7.138405    7.807623   10.655958   

                                         2001     ...            2008  \
annee_permis   annee_permis_index                 ...                   
16 ans et plus 16.0                170.031310     ...      139.520963   
0 an           0.0                 158.154178     ...      103.349804   
1 an           1.0                 121.968624     ...       75.841267   
2 ans          2.0                  57.694045     ...       42.361811   
3 ans          3.0                  17.423118     ...       12.650185   
4 ans          4.0                  15.944966     ...       11.496095   
5 ans          5.0                  13.612848     ...        9.823983   
6-10 ans       6.0                  11.006537     ...        7.497953   
11-15 ans      11.0                  7.308612     ...        4.576432   
Inconnu        999.0                 9.127786     ...        9.005324   

                                         2009        2010        2011  \
annee_permis   annee_permis_index                                       
16 ans et plus 16.0                138.460565  132.426681  124.912390   
0 an           0.0                 107.369984   91.846562   90.650620   
1 an           1.0                  69.190132   60.503807   59.967035   
2 ans          2.0                  39.063671   36.455055   30.937408   
3 ans          3.0                  12.532685    9.571983    9.519724   
4 ans          4.0    

> Note that for precise period sush as "6-10 ans" and "11-15 ans" results are divide by the number of years. Since we have no information about the driver licence in the population, we can't report by the number of citizen who have a driver licence.

### C.3. Sexe and Age of drivers excluding young drivers (having driver licence for 5 years or more)

In [350]:
df_conducteur_confirmed = df_Accident_objets_normalized[(df_Accident_objets_normalized.faute_index==1)\
                                           & (df_Accident_objets_normalized.voiture_index==210)\
                                            &(df_Accident_objets_normalized.annee_permis_index>4)\
                                             &(df_Accident_objets_normalized.annee_permis_index!=999)]
df_conducteur_confirmed = df_conducteur_confirmed.groupby(['sexe','age']).sum().sort_values(['2017'], ascending=False)[[col for col in df_conducteur.columns if IS_DATE.match(col)]]
display(df_conducteur_confirmed.head(5))


1992       1993       1994       1995       1996  \
sexe  age                                                                     
Homme 70 ans et plus  26.605339  27.894003  27.603285  28.914694  27.268483   
      30-39 ans       36.774496  36.132917  36.288853  35.853885  34.970209   
      60-69 ans       29.720316  29.456963  28.886025  28.985116  27.173187   
      40-49 ans       30.786296  31.017370  31.461494  31.431934  29.587514   
      50-59 ans       30.803029  31.345814  31.282173  30.602205  27.632502   

                           1997       1998       1999       2000       2001  \
sexe  age                                                                     
Homme 70 ans et plus  28.520126  25.248015  26.284672  30.398260  28.809353   
      30-39 ans       35.348139  34.494125  39.467690  36.208119  37.701855   
      60-69 ans       27.987325  26.993640  26.403598  28.517504  26.699334   
      40-49 ans       28.997091  32.191127  30.611140  32.486644  32.811370   
      50-59 ans       29.429806  28.165035  28.889887  30.956427  27.969640   

                        ...           2008       2009       2010       2011  \
sexe  age               ...                                                   
Homme 70 ans et plus    ...      22.113104  22.846748  21.197766  21.546216   
      30-39 ans         ...      25.810091  26.045067  24.134367  21.764802   
      60-69 ans         ...      20.450934  21.256262  20.120053  19.172347   
      40-49 ans         ...      24.490993  23.119928  22.730356  19.868757   
      50-59 ans         ...      23.222267  21.855012  21.722033  19.558043   

                           2012       2013       2014       2015       2016  \
sexe  age                                                                     
Homme 70 ans et plus  19.759087  20.454627  18.856224  19.972230  20.870296   
      30-39 ans       21.154606  20.895599  19.757757  17.507229  17.939887   
      60-69 ans       17.843825  18.024530  18.112347  16.667834  17.176259   
      40-49 ans       20.444228  18.909912  17.964703  16.953680  16.877826   
      50-59 ans       19.001292  17.514302  18.078431  17.922416  17.134273   

                           2017  
sexe  age                        
Homme 70 ans et plus  19.699791  
      30-39 ans       18.679447  
      60-69 ans       16.579135  
      40-49 ans       16.552956  
      50-59 ans       16.080889  

[5 rows x 26 columns]

> Clearly **today**, old men are dangerous!

### C.4. Sexe and Age of conductors of young drivers (having driver licence for 4 years or less)

In [351]:
df_conducteur_young = df_Accident_objets_normalized[(df_Accident_objets_normalized.faute_index==1)\
                                           & (df_Accident_objets_normalized.voiture_index==210)\
                                            &(df_Accident_objets_normalized.annee_permis_index<5)\
                                             &(df_Accident_objets_normalized.annee_permis_index!=999)]
df_conducteur_young = df_conducteur_young.groupby(['sexe','age']).sum().sort_values(['2017'], ascending=False)[[col for col in df_conducteur.columns if IS_DATE.match(col)]]
display(df_conducteur_young.head(5))

1992        1993        1994        1995       1996  \
sexe  age                                                                    
Homme 20 ans     123.945930  113.200860  120.586887  111.194697  89.596213   
      18-19 ans   86.830022   81.095017   75.234055   75.854428  80.036059   
Femme 20 ans      40.550263   46.989051   39.646953   41.184166  36.045823   
      18-19 ans   26.874978   24.238099   24.733518   21.767192  25.926588   
Homme 21-29 ans   40.718091   37.982672   38.267575   37.628218  32.465106   

                       1997        1998        1999        2000        2001  \
sexe  age                                                                     
Homme 20 ans     106.085985  113.196594  127.281460  111.513735  126.854243   
      18-19 ans   75.322463   82.696359   84.615294   93.176688   95.369426   
Femme 20 ans      33.162947   37.554189   39.472369   47.095380   47.158633   
      18-19 ans   19.355500   25.117066   26.732342   25.583274   29.917636   
Homme 21-29 ans   33.678133   37.736919   40.237612   39.439721   38.065340   

                   ...           2008       2009       2010       2011  \
sexe  age          ...                                                   
Homme 20 ans       ...      74.935784  67.243634  58.860037  59.317158   
      18-19 ans    ...      65.759517  63.028279  54.045739  59.131013   
Femme 20 ans       ...      33.497884  37.035442  32.483758  24.248202   
      18-19 ans    ...      20.038072  22.608543  21.058965  18.344617   
Homme 21-29 ans    ...      26.740257  26.234701  22.062705  19.590199   

                      2012       2013       2014       2015       2016  \
sexe  age                                                                
Homme 20 ans     53.885762  44.372648  46.859014  39.032480  43.152761   
      18-19 ans  48.573933  45.988975  42.696748  37.698823  36.461752   
Femme 20 ans     27.080663  25.789126  30.723616  26.294974  24.921720   
      18-19 ans  21.186913  21.355098  17.147693  16.814770  18.416622   
Homme 21-29 ans  19.174300  17.928773  16.632623  15.226247  15.647355   

                      2017  
sexe  age                   
Homme 20 ans     39.736442  
      18-19 ans  34.914941  
Femme 20 ans     25.314246  
      18-19 ans  18.651166  
Homme 21-29 ans  14.362600  

[5 rows x 26 columns]

### C.5. Who has not the driver licence

In [352]:
df_conducteur_illegal = df_Accident_objets_normalized[(df_Accident_objets_normalized.faute_index==1)\
                                           & (df_Accident_objets_normalized.voiture_index==210)\
                                             &(df_Accident_objets_normalized.annee_permis_index==999)]
df_conducteur_illegal = df_conducteur_illegal.groupby(['sexe','age']).sum().sort_values(['2017'], ascending=False)[[col for col in df_conducteur.columns if IS_DATE.match(col)]]
display(df_conducteur_illegal.head(5))

1992      1993      1994      1995      1996      1997  \
sexe  age                                                                     
Homme 15-17 ans  1.824409  1.498339  1.408392  1.471093  1.528806  1.111588   
      21-29 ans  1.641859  1.309747  1.103872  1.056379  0.901198  1.030963   
      18-19 ans  2.045734  1.640651  2.043097  1.582837  1.461846  1.703578   
      30-39 ans  0.508080  0.447786  0.556733  0.497510  0.542556  0.362207   
Femme 18-19 ans  0.238889  0.000000  0.000000  0.632767  0.127717  0.128182   

                     1998      1999      2000      2001    ...         2008  \
sexe  age                                                  ...                
Homme 15-17 ans  1.419111  1.567521  1.867966  1.484514    ...     1.419930   
      21-29 ans  1.086975  1.399989  1.370114  1.296046    ...     1.036269   
      18-19 ans  1.685224  1.292734  1.851959  1.854084    ...     2.541431   
      30-39 ans  0.296653  0.561230  0.612855  0.425086    ...     0.517630   
Femme 18-19 ans  0.252433  0.247522  0.000000  0.363372    ...     0.111323   

                     2009      2010      2011      2012      2013      2014  \
sexe  age                                                                     
Homme 15-17 ans  1.148493  1.033928  1.109287  1.487144  0.671296  0.451169   
      21-29 ans  1.238738  0.757157  0.806777  0.694277  0.766359  0.698850   
      18-19 ans  2.087029  1.640857  0.831367  1.267146  1.062101  0.632544   
      30-39 ans  0.286210  0.481973  0.335386  0.417869  0.256913  0.218783   
Femme 18-19 ans  0.000000  0.107444  0.000000  0.223020  0.223614  0.000000   

                     2015      2016      2017  
sexe  age                                      
Homme 15-17 ans  0.593331  0.525782  0.534870  
      21-29 ans  0.393952  0.335019  0.355119  
      18-19 ans  0.206569  1.354294  0.318373  
      30-39 ans  0.247978  0.228742  0.258539  
Femme 18-19 ans  0.111356  0.000000  0.228849  

[5 rows x 26 columns]

## D. Analyse on the actor of car accident in proportion of population who have the driver licence

### D.0. Normalize the data by population with driver licence possession

In [356]:
df_conducteur_bysexe = df_Population_2010[(df_Population_2010['canton'] == 'Suisse') & (df_Population_2010['sexe'] != 'Sexe - Total')\
                                     & (df_Population_2010.etat_civil == "Etat civil - Total")]\
                .drop(columns=['canton','canton_index','permanente','etat_civil','Age - Total'])
df_conducteur_bysexe['annee'] = df_conducteur_bysexe['annee'].map(int).map(str)
df_conducteur_bysexe = df_conducteur_bysexe.groupby(['annee','sexe']).sum().transpose().sum()
df_conducteur_bysexe = pd.DataFrame(df_conducteur_bysexe)

df_conducteur_bysexe = df_conducteur_bysexe.transpose()
df_conducteur_1992 = df_Population_age_1992[(df_Population_age_1992.age == "Age - Total") & (df_Population_age_1992.sexe != 'Sexe - Total')]\
                        .drop(columns=['age'])
for year in range(1992,2010):
    for sexe in np.unique([sexe[1] for sexe in df_conducteur_bysexe.columns.values]):
        df_conducteur_bysexe[(str(year),sexe)] = df_conducteur_1992.loc[df_conducteur_1992.sexe==sexe][str(year)].values[0]
df_conducteur_bysexe = df_conducteur_bysexe.transpose().reset_index()
annee = []
for elem in df_proportion_permis.columns:
    if IS_DATE.match(elem):
        annee.append(elem)
df_conducteur_bysexe = df_conducteur_bysexe[df_conducteur_bysexe.annee.isin(annee)].set_index(['annee','sexe']).transpose()
df_proportion_permis_sexe = df_proportion_permis[df_proportion_permis.critere.isin(['Hommes','Femmes'])]
for year in annee:
    for sexe in np.unique([sexe[1] for sexe in df_conducteur_bysexe.columns.values]):
        df_conducteur_bysexe[(str(year),sexe)] = int((1/100) * df_conducteur_bysexe[(str(year),sexe)].values[0] * \
                                            df_proportion_permis_sexe.loc[df_proportion_permis_sexe.critere==sexe+'s'][str(year)].values[0])
df_conducteur_bysexe.head(5)

annee     2010              2015              1994              2000           \
sexe     Femme    Homme    Femme    Homme    Femme    Homme    Femme    Homme   
0      5982246  6886822  6407827  7358195  2377216  2989453  2616965  3125519   

annee     2005           
sexe     Femme    Homme  
0      2841956  3283894

In [359]:
toKeep = []
for col in df_Accident_objets.columns:
    if IS_DATE.match(col):
        if(col in annee):
            toKeep.append(col)
    else :
        toKeep.append(col)

In [360]:
df_Accident_objets_normalized_sexe = df_Accident_objets[toKeep]

In [361]:
for year,sexe in df_conducteur_bysexe.columns.values:
        df_Accident_objets_normalized_sexe.loc[(df_Accident_objets_normalized_sexe.sexe==sexe),year]\
        = df_Accident_objets_normalized_sexe[df_Accident_objets_normalized_sexe.sexe==sexe][year]*10000 / df_conducteur_bysexe.transpose().loc[(year,sexe)].values[0]

C:\Users\Clementg20\Anaconda3\lib\site-packages\pandas\core\indexing.py:543: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [362]:
df_conducteur = df_Accident_objets_normalized_sexe[(df_Accident_objets_normalized_sexe.faute_index==1)\
                                           & (df_Accident_objets_normalized_sexe.voiture_index==210)\
                                            & (df_Accident_objets_normalized_sexe.sexe_index <= 2)].dropna()

In [363]:
for i in df_conducteur['1994']:
    if type(i) != float:
        print(type(i))

In [364]:
display(df_conducteur.groupby(['sexe']).sum())

voiture_index  age_index  sexe_index  annee_permis_index  faute_index  \
sexe                                                                           
Femme        88200.0    40320.0       840.0             43974.0        420.0   
Homme        88200.0    40320.0       420.0             43974.0        420.0   

       gravite_index       1994       2000       2005       2010      2015  
sexe                                                                        
Femme       132720.0  20.242166  20.145474  17.871494   7.913416  6.506106  
Homme       132720.0  40.706444  38.940733  31.203809  12.921780  9.813548

In [314]:
df_proportion_permis

critere     1994     2000     2005     2010     2015
0           Total  76.1925  79.5684  81.4970  80.8099  81.5869
1          Hommes  86.7108  88.6161  89.1752  87.8091  87.9242
2          Femmes  66.2318  71.0744  74.4214  74.4154  75.5780
3       18-24 ans  70.7393  68.7840  62.7032  58.6586  60.9546
4       25-64 ans  84.8314  87.8657  89.5716  89.3944  88.8073
5  65 ans et plus  44.2482  53.2861  61.7433  64.0817  69.2718

In [366]:
df_Accident_cantons.groupby(['canton']).sum()

canton_index  gravite_index  ruralite_index  \
canton                                                                          
Aargau                                   171.0         2844.0            99.0   
Appenzell Ausserrhoden                   135.0         2844.0            99.0   
Appenzell Innerrhoden                    144.0         2844.0            99.0   
Basel-Landschaft                         117.0         2844.0            99.0   
Basel-Stadt                              108.0         2844.0            99.0   
Bern / Berne                              18.0         2844.0            99.0   
Fribourg / Freiburg                       90.0         2844.0            99.0   
Genève                                   225.0         2844.0            99.0   
Glarus                                    72.0         2844.0            99.0   
Graubünden / Grigioni / Grischun         162.0         2844.0            99.0   
Jura                                     260.0         3161.0           111.0   
Luzern                                    27.0         2844.0            99.0   
Neuchâtel                                216.0         2844.0            99.0   
Nidwalden                                 63.0         2844.0            99.0   
Obwalden                                  54.0         2844.0            99.0   
Schaffhausen                             126.0         2844.0            99.0   
Schwyz                                    45.0         2844.0            99.0   
Solothurn                                 99.0         2844.0            99.0   
St. Gallen                               153.0         2844.0            99.0   
Thurgau                                  180.0         2844.0            99.0   
Ticino                                   189.0         2844.0            99.0   
Uri                                       36.0         2844.0            99.0   
Valais / Wallis                          207.0         2844.0            99.0   
Vaud                                     198.0         2844.0            99.0   
Zug                                       81.0         2844.0            99.0   
Zürich                                     9.0         2844.0            99.0   

                                    1992    1993    1994    1995    1996  \
canton                                                                     
Aargau                            1547.0  1531.0  1691.0  1576.0  1544.0   
Appenzell Ausserrhoden             132.0   138.0   135.0   127.0   113.0   
Appenzell Innerrhoden               32.0    30.0    29.0    38.0    27.0   
Basel-Landschaft                   574.0   552.0   674.0   724.0   636.0   
Basel-Stadt                        732.0   787.0   864.0   858.0   785.0   
Bern / Berne                      3237.0  3176.0  3234.0  3176.0  3019.0   
Fribourg / Freiburg                600.0   573.0   638.0   638.0   554.0   
Genève                            1122.0  1084.0  1125.0  1039.0  1033.0   
Glarus                             110.0    98.0   110.0    81.0    88.0   
Graubünden / Grigioni / Grischun   819.0   837.0   786.0   778.0   748.0   
Jura                               224.0   247.0   197.0   218.0   212.0   
Luzern                            1105.0  1058.0  1057.0  1079.0  1034.0   
Neuchâtel                          569.0   518.0   540.0   485.0   470.0   
Nidwalden                          103.0    82.0   113.0   108.0   113.0   
Obwalden                           113.0   110.0   100.0    94.0    96.0   
Schaffhausen                       168.0   151.0   197.0   180.0   173.0   
Schwyz                             416.0   374.0   415.0   467.0   397.0   
Solothurn                          767.0   777.0   813.0   799.0   720.0   
St. Gallen                        1310.0  1322.0  1349.0  1263.0  1179.0   
Thurgau                            682.0   738.0   659.0   612.0   570.0   
Ticino                            1771.0  1723.0  1762.0  1708.0  1641.0   
Uri              

In [367]:
describeDf(df_Accident_cantons)

_________________ DESCRIPTION _________________


         ATTRIBUTE: canton_index
   [ 1.  2.  3.  4.  5.  6.  7.  8.  9. 10. 11. 12. 13. 14. 15. 16. 17. 18.
 19. 20. 21. 22. 23. 24. 25. 26.]


         ATTRIBUTE: canton
   ['Zürich' 'Bern / Berne' 'Luzern' 'Uri' 'Schwyz' 'Obwalden' 'Nidwalden'
 'Glarus' 'Zug' 'Fribourg / Freiburg' 'Solothurn' 'Basel-Stadt'
 'Basel-Landschaft' 'Schaffhausen' 'Appenzell Ausserrhoden'
 'Appenzell Innerrhoden' 'St. Gallen' 'Graubünden / Grigioni / Grischun'
 'Aargau' 'Thurgau' 'Ticino' 'Vaud' 'Valais / Wallis' 'Neuchâtel' 'Genève'
 'Jura']


         ATTRIBUTE: gravite_index
   [315. 316. 317.]


         ATTRIBUTE: gravite
   ['Accidents avec blessés légers' 'Accidents avec blessés graves'
 'Accidents mortels']


         ATTRIBUTE: ruralite_index
   [10. 11. 12.]


         ATTRIBUTE: ruralite
   ['En localité' 'Hors localité' 'Sur autoroute']


         VALUES: ['1992', '1993', '1994', '1995', '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2